File: python/src/iskra/gh.py
# src/iskra/gh.py
import sys
import subprocess
import argparse
import json
import webbrowser

from rich.console import Console
from rich.table import Table

from rich.panel import Panel
from .github import get_prs

from iskra.ui.formatting import get_icon

console = Console()


def get_git_root(path: str = ".") -> str | None:
    result = subprocess.run(
        ["git", "-C", path, "rev-parse", "--show-toplevel"],
        capture_output=True,
        text=True,
    )
    if result.returncode != 0:
        return None
    return result.stdout.strip()


def get_remote_url(repo_path: str) -> str | None:
    result = subprocess.run(
        ["git", "-C", repo_path, "remote", "get-url", "origin"],
        capture_output=True,
        text=True,
    )
    if result.returncode != 0:
        return None
    return result.stdout.strip()


def parse_github_slug(remote_url: str) -> str | None:
    # supports:
    #   git@github.com:user/repo.git
    #   https://github.com/user/repo.git
    #   https://github.com/user/repo
    if "github.com" not in remote_url:
        return None

    url = remote_url.rstrip("/")
    url = url.removesuffix(".git")

    if url.startswith("git@github.com:"):
        return url.split("git@github.com:")[1]
    if "github.com/" in url:
        return url.split("github.com/")[1]

    return None


def cmd_info(repo_path: str) -> int:
    remote = get_remote_url(repo_path)
    if not remote:
        console.print("[red]No 'origin' remote found[/]")
        return 1

    slug = parse_github_slug(remote)

    table = Table(
        title=f"{get_icon('project')} GitHub Info",
        show_header=False,
        box=None,
        padding=(0, 2),
    )
    table.add_column("Field", style="cyan")
    table.add_column("Value", style="white")

    table.add_row("Repo path", repo_path)
    table.add_row("Remote", remote)
    table.add_row("GitHub slug", slug or "[dim]not a GitHub repo[/]")

    console.print(table)
    return 0


def cmd_open(repo_path: str) -> int:
    remote = get_remote_url(repo_path)
    if not remote:
        console.print("[red]No 'origin' remote found[/]")
        return 1

    slug = parse_github_slug(remote)
    if not slug:
        console.print("[red]Remote is not a GitHub URL[/]")
        return 1

    url = f"https://github.com/{slug}"
    console.print(f"{get_icon('link')} Opening [bold]{url}[/]...")
    webbrowser.open(url)
    return 0


def cmd_prs(
    repo_path: str,
    limit: int,
    state: str,
    draft: str,
    need_review: bool,
    require_changes: bool,
    open_number: int | None,
) -> int:
    remote = get_remote_url(repo_path)
    if not remote:
        console.print("[red]No 'origin' remote found[/]")
        return 1

    slug = parse_github_slug(remote)
    if not slug:
        console.print("[red]Remote is not a GitHub URL[/]")
        return 1

    try:
        console.print(
            f"[bold blue]{
                get_icon('github')
            } Fetching PRs for {slug}...[/]"
        )
        prs = get_prs(
            slug=slug,
            limit=limit,
            state=state,
            draft=draft,
            need_review=need_review,
            require_changes=require_changes,
        )
    except subprocess.CalledProcessError as e:
        console.print(
            f"[bold red]{get_icon('error')} Error fetching PRs from GitHub:[/]"
        )
        console.print(
            Panel(
                e.stderr or str(e),
                title="Error Details",
                border_style="red",
            )
        )
        return 1
    except json.JSONDecodeError as e:
        console.print(
            f"[bold red]{get_icon('error')} Error parsing GitHub response:[/]"
        )
        console.print(Panel(str(e), title="JSON Error", border_style="red"))
        return 1

    if open_number is not None:
        match = next(
            (
                pr
                for pr in prs
                if pr.get(
                    "number",
                )
                == open_number
            ),
            None,
        )
        if not match:
            console.print(
                f"[red]PR #{
                    open_number
                } not found in listed results[/]"
            )
            return 1
        url = match.get("url")
        if not url:
            console.print("[red]PR has no URL[/]")
            return 1
        console.print(f"{get_icon('link')} Opening [bold]{url}[/]...")
        webbrowser.open(url)
        return 0

    table = Table(
        title=f"{get_icon('project')} Pull Requests ({slug})",
        show_header=True,
        box=None,
        padding=(0, 1),
    )
    table.add_column("#", style="cyan", justify="right")
    table.add_column("Title", style="white")
    table.add_column("State", style="magenta")
    table.add_column("Draft", style="dim")
    table.add_column("Review", style="yellow")

    for pr in prs:
        table.add_row(
            str(pr.get("number", "")),
            pr.get("title", "") or "",
            pr.get("state", "") or "",
            "yes" if pr.get("isDraft", False) else "",
            pr.get("reviewDecision") or "",
        )

    console.print(table)
    return 0


def main(argv: list[str] | None = None) -> int:
    if argv is None:
        argv = sys.argv[1:]

    parser = argparse.ArgumentParser(
        description="Iskra GitHub integration",
        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
    )

    subparsers = parser.add_subparsers(dest="command", required=True)

    subparsers.add_parser("info", help="Show GitHub info for current repo")
    subparsers.add_parser("open", help="Open GitHub repo page in browser")

    # later: issues, prs, status, etc.

    prs_parser = subparsers.add_parser(
        "prs", help="List pull requests for current repo"
    )
    prs_parser.add_argument("--limit", type=int, default=50)
    prs_parser.add_argument(
        "--state",
        type=str,
        default="open",
        choices=[
            "open",
            "closed",
            "merged",
            "all",
        ],
    )
    prs_parser.add_argument(
        "--draft", type=str, default="all", choices=["all", "only", "exclude"]
    )
    prs_parser.add_argument("--need-review", action="store_true")
    prs_parser.add_argument("--require-changes", action="store_true")
    prs_parser.add_argument(
        "--open", type=int, default=None, help="Open PR number in browser"
    )
    args = parser.parse_args(argv)

    repo_root = get_git_root(".")
    if not repo_root:
        console.print("[red]iskra gh: not inside a git repository[/]")
        return 1

    if args.command == "info":
        return cmd_info(repo_root)
    if args.command == "open":
        return cmd_open(repo_root)
    if args.command == "prs":
        return cmd_prs(
            repo_root,
            limit=args.limit,
            state=args.state,
            draft=args.draft,
            need_review=args.need_review,
            require_changes=args.require_changes,
            open_number=args.open,
        )

    console.print("[red]Unknown gh command[/]")
    return 1

File: go-core/internal/gh/gh.go
// Package gh implements iskra gh subcommands: info, open, prs.
// All GitHub API calls go through the gh CLI.
package gh

import (
	"encoding/json"
	"fmt"
	"os/exec"
	"strings"

	"github.com/NoamFav/iskra/internal/ui"
)

// GitRoot returns the git root of the given path, or "" if not a git repo.
func GitRoot(path string) string {
	out, err := exec.Command("git", "-C", path, "rev-parse", "--show-toplevel").Output()
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(out))
}

// RemoteURL returns the origin remote URL, or "" if none.
func RemoteURL(repoPath string) string {
	out, err := exec.Command("git", "-C", repoPath, "remote", "get-url", "origin").Output()
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(out))
}

// GithubSlug parses a git remote URL into owner/repo form, or "" if not GitHub.
func GithubSlug(remoteURL string) string {
	if !strings.Contains(remoteURL, "github.com") {
		return ""
	}
	url := strings.TrimRight(remoteURL, "/")
	url = strings.TrimSuffix(url, ".git")
	if strings.HasPrefix(url, "git@github.com:") {
		return strings.TrimPrefix(url, "git@github.com:")
	}
	if idx := strings.Index(url, "github.com/"); idx >= 0 {
		return url[idx+len("github.com/"):]
	}
	return ""
}

// RunInfo prints GitHub info for the current repo.
func RunInfo(repoPath string) int {
	remote := RemoteURL(repoPath)
	if remote == "" {
		ui.ErrorMsg("No 'origin' remote found")
		return 1
	}
	slug := GithubSlug(remote)
	slugStr := slug
	if slugStr == "" {
		slugStr = ui.Mute("not a GitHub repo")
	}

	fmt.Printf("\n  %-12s %s\n", ui.Bold("Path"), repoPath)
	fmt.Printf("  %-12s %s\n", ui.Bold("Remote"), remote)
	fmt.Printf("  %-12s %s\n\n", ui.Bold("Slug"), slugStr)
	return 0
}

// RunOpen opens the GitHub repo page in the default browser.
func RunOpen(repoPath string) int {
	remote := RemoteURL(repoPath)
	if remote == "" {
		ui.ErrorMsg("No 'origin' remote found")
		return 1
	}
	slug := GithubSlug(remote)
	if slug == "" {
		ui.ErrorMsg("Remote is not a GitHub URL")
		return 1
	}
	url := "https://github.com/" + slug
	fmt.Printf("%s Opening %s\n", ui.Icons.Arrow, ui.Inf(url))
	if err := exec.Command("open", url).Start(); err != nil {
		// fallback for Linux
		exec.Command("xdg-open", url).Start()
	}
	return 0
}

// PR holds fields from `gh pr list --json`.
type PR struct {
	Number         int    `json:"number"`
	Title          string `json:"title"`
	State          string `json:"state"`
	IsDraft        bool   `json:"isDraft"`
	URL            string `json:"url"`
	ReviewDecision string `json:"reviewDecision"`
	Author         struct {
		Login string `json:"login"`
	} `json:"author"`
	HeadRefName string `json:"headRefName"`
	BaseRefName string `json:"baseRefName"`
}

// RunPRs lists or opens a pull request for the current repo.
func RunPRs(repoPath string, limit int, state, draft string, needReview, requireChanges bool, openNumber int) int {
	remote := RemoteURL(repoPath)
	if remote == "" {
		ui.ErrorMsg("No 'origin' remote found")
		return 1
	}
	slug := GithubSlug(remote)
	if slug == "" {
		ui.ErrorMsg("Remote is not a GitHub URL")
		return 1
	}

	fields := "number,title,state,isDraft,url,author,createdAt,reviewDecision,headRefName,baseRefName"
	cmd := exec.Command("gh", "pr", "list",
		"--repo", slug,
		"--limit", fmt.Sprintf("%d", limit),
		"--state", state,
		"--json", fields,
	)

	fmt.Printf("%s Fetching PRs for %s...\n", ui.Icons.Git, ui.Bold(slug))

	out, err := cmd.Output()
	if err != nil {
		ui.ErrorMsg("gh pr list failed: " + err.Error())
		return 1
	}

	var prs []PR
	if err := json.Unmarshal(out, &prs); err != nil {
		ui.ErrorMsg("Failed to parse gh output: " + err.Error())
		return 1
	}

	// Apply filters
	var filtered []PR
	for _, pr := range prs {
		if draft == "only" && !pr.IsDraft {
			continue
		}
		if draft == "exclude" && pr.IsDraft {
			continue
		}
		if needReview && pr.ReviewDecision != "" && pr.ReviewDecision != "REVIEW_REQUIRED" {
			continue
		}
		if requireChanges && pr.ReviewDecision != "CHANGES_REQUESTED" {
			continue
		}
		filtered = append(filtered, pr)
	}

	// Open a specific PR in browser
	if openNumber > 0 {
		for _, pr := range filtered {
			if pr.Number == openNumber {
				fmt.Printf("%s Opening %s\n", ui.Icons.Arrow, ui.Inf(pr.URL))
				if err := exec.Command("open", pr.URL).Start(); err != nil {
					exec.Command("xdg-open", pr.URL).Start()
				}
				return 0
			}
		}
		ui.ErrorMsg(fmt.Sprintf("PR #%d not found", openNumber))
		return 1
	}

	if len(filtered) == 0 {
		ui.WarningMsg("No pull requests found")
		return 0
	}

	// Print table
	fmt.Printf("\n  %s Pull Requests — %s (%d)\n\n", ui.Icons.Git, ui.Bold(slug), len(filtered))
	fmt.Printf("  %-6s  %-50s  %-8s  %-5s  %s\n",
		ui.Bold("#"), ui.Bold("Title"), ui.Bold("State"), ui.Bold("Draft"), ui.Bold("Review"))
	fmt.Printf("  %s\n", strings.Repeat("─", 85))

	for _, pr := range filtered {
		draft := ""
		if pr.IsDraft {
			draft = "yes"
		}
		title := pr.Title
		if len(title) > 50 {
			title = title[:47] + "..."
		}
		stateStr := pr.State
		switch strings.ToLower(pr.State) {
		case "open":
			stateStr = ui.Success(pr.State)
		case "closed":
			stateStr = ui.Mute(pr.State)
		case "merged":
			stateStr = ui.Inf(pr.State)
		}
		fmt.Printf("  %-6d  %-50s  %-8s  %-5s  %s\n",
			pr.Number, title, stateStr, draft, pr.ReviewDecision)
	}
	fmt.Println()
	return 0
}
